src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2653,269 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
+ * generated columns cache from the catalog.
+ */
+static bool
+generated_column_iter_hook(void *ctx, GeneratedColumn *column)
+{
+	StreamSpecs *specs = (StreamSpecs *) ctx;
+	StreamContext *privateContext = &(specs->private);
+
+	GeneratedColumnsCache *item = (GeneratedColumnsCache *)
+								  calloc(1, sizeof(GeneratedColumnsCache));
+
+	if (item == NULL)
+	{
+		log_error(ALLOCATION_FAILED_ERROR);
+		return false;
+	}
+
+	(void) normalizedGeneratedColumn(&(item->generatedColumn),
+									 column->nspname,
+									 column->relname,
+									 column->attname);
+
+	HASH_ADD(hh,
+			 privateContext->generatedColumnsCache,
+			 generatedColumn,
+			 sizeof(GeneratedColumn),
+			 item);
+
+	return true;
+}
+
+
+/*
+ * prepareGeneratedColumnsCache fills-in the cache with the tables having
+ * generated columns.
+ */
+static bool
+prepareGeneratedColumnsCache(StreamSpecs *specs)
+{
+	/*
+	 * TODO: GeneratedColumn must be retrieved from the target catalog
+	 * because the schema of the target can be different from the source.
+	 */
+	if (!catalog_iter_s_generated_column(specs->sourceDB,
+										 specs,
+										 &generated_column_iter_hook))
+	{
+		log_error("Failed to prepare a generated column cache for our catalog,"
+				  "see above for details");
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * removeGeneratedColumnsFromTransaction removes the generated columns from the
+ * given transaction.
+ */
+static bool
+removeGeneratedColumnsFromTransaction(GeneratedColumnsCache *cache,
+									  LogicalTransaction *txn)
+{
+	LogicalTransactionStatement *stmt = txn->first;
+
+	for (; stmt != NULL; stmt = stmt->next)
+	{
+		if (!removeGeneratedColumnsFromStatement(cache, stmt))
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * removeGeneratedColumnsFromStatement removes the generated columns from the
+ * given statement after looking up the cache.
+ */
+static bool
+removeGeneratedColumnsFromStatement(GeneratedColumnsCache *cache,
+									LogicalTransactionStatement *stmt)
+{
+	LogicalMessageTupleArray *columns = NULL;
+	const char *nspname = NULL;
+	const char *relname = NULL;
+
+	if (stmt->action == STREAM_ACTION_INSERT)
+	{
+		columns = &(stmt->stmt.insert.new);
+		nspname = stmt->stmt.insert.table.nspname;
+		relname = stmt->stmt.insert.table.relname;
+	}
+	else if (stmt->action == STREAM_ACTION_UPDATE)
+	{
+		columns = &(stmt->stmt.update.new);
+		nspname = stmt->stmt.update.table.nspname;
+		relname = stmt->stmt.update.table.relname;
+	}
+	else
+	{
+		/* Only INSERT and UPDATE statements might have generated columns */
+		return true;
+	}
+
+	if (columns->count > 1)
+	{
+		log_error("BUG: removeGeneratedColumnsFromStatement called on a statement "
+				  "with multiple VALUES clauses");
+		return false;
+	}
+
+	LogicalMessageTuple *tuple = &(columns->array[0]);
+
+	for (int c = 0; c < tuple->cols; c++)
+	{
+		if (isGeneratedColumn(cache,
+							  nspname,
+							  relname,
+							  tuple->columns[c]))
+		{
```

Comment:
Is there a way to bypass walking through the list of attribute of a table for every INSERT and UPDATE operation we want to transform? Maybe another level of caching to know if a table has any generated columns?